lms/djangoapps/completion_api/models.py

```diff
@@ -0,0 +1,276 @@
+
+    @property
+    def earned(self):
+        """
+        Return the number of completions earned by the user.
+        """
+        return self._inner.completions
+
+    def iter_block_keys_in_category(self, category):
+        """
+        Yields the UsageKey for all blocks of the specified category.
+        """
+        for block in self.blocks:
+            if self.blocks.get_xblock_field(block, 'category') == category:
+                yield block
+
+    def get_completions_in_category(self, category):
+        """
+        Returns a list of BlockCompletions for each block of the requested category.
+        """
+        if category not in self._completions_in_category:
+            completions = [
+                BlockCompletion(self.user, block_key, self) for block_key in self.iter_block_keys_in_category(category)
+            ]
+            self._completions_in_category[category] = completions
+        return self._completions_in_category[category]
+
+    @property
+    def chapter(self):
+        """
+        Return a list of BlockCompletions for each chapter in the course.
+        """
+        return self.get_completions_in_category('chapter')
+
+    @property
+    def sequential(self):
+        """
+        Return a list of BlockCompletions for each sequential in the course.
+        """
+        return self.get_completions_in_category('sequential')
+
+    @property
+    def vertical(self):
+        """
+        Return a list of BlockCompletions for each vertical in the course.
+        """
+        return self.get_completions_in_category('vertical')
+
+
+class BlockCompletion(CompletionDataMixin, object):
+    """
+    Class to represent completed blocks within a given block of the course.
+    """
+
+    def __init__(self, user, block_key, course_completion):
+        self.user = user
+        self.block_key = block_key
+        self.course_key = block_key.course_key
+        self.course_completion = course_completion
+        self._blocks = None
+        self._completable_blocks = None
+
+    @property
+    def blocks(self):
+        """
+        Return an `openedx.core.lib.block_structure.block_structure.BlockStructureBlockData`
+        object which behaves as dict that maps `opaque_keys.edx.locator.BlockUsageLocator`s
+        to `openedx.core.lib.block_structure.block_structure.BlockData` objects
+        for all blocks in the sub-tree (or DAG) under self.block_key.
+        """
+        if self._blocks is None:
+
+            self._blocks = get_course_blocks(
+                self.user,
+                self.block_key,
+                collected_block_structure=self.course_completion.collected,
+            )
+        return self._blocks
+
+    @property
+    def completed_blocks(self):
+        """
+        Return the list of UsageKeys of all blocks within self.block that have been
+        completed by self.user.
+        """
+        modules = CourseModuleCompletion.objects.filter(
```

Comment:
Should this query be cached on the `self.course_completion` object as well? It's not specific to any subtree.

Reply:
Also a plausible source of slowdown.  I'll fix it.